django/core/management/commands/makemessages.py

```diff
@@ -427,12 +431,23 @@ def handle(self, *args, **options):
 
             # Build po files for each selected locale
             for locale in locales:
-                if "-" in locale:
+                if not is_valid_locale(locale):
+                    if "_" in locale:
+                        locale_parts = locale.partition("_")
+                    else:
+                        locale_parts = locale.partition("-")
+                    proposed_locale = "".join(
+                        (
+                            locale_parts[0].lower(),
+                            locale_parts[1].replace("-", "_"),
+                            locale_parts[2][:2].upper() + locale_parts[2][2:],
+                        )
+                    )
```

Comment:
This block looks under-tests. (Is there a check for the ` locale_parts[2][:2].upper() + locale_parts[2][2:]` part of `proposed_locale` for instance? 🤔)

I think the comment here could be expanded. Maybe not enough :coffee: yet but, the logic here seems quite opaque. Is someone (me?) coming to this in 18months time going to see what's in play here? Not sure…

Reply:
As far as I can see this block is not covered by the tests. (An `assert False` on line 435 is never hit.) 

We need a regression test to confirm the behaviour being added here — at the very least to make sure we don't break it going forward.

Reply:
I gave the code another look, and indeed it was a bit opaque. I've rewritten it and added extra testcases.